main.py

```diff
@@ -67,11 +68,71 @@ def request_sample(dev: serial.Serial) -> str:
 }
 
 
+def start_publishing(args, plugin, dev):
+    """
+    start_publishing initializes the raingauge and begins sampling and publishing data
+    """
+    # initialize the raingauge
+    try:
+        logging.info("set to polling mode")
+        set_mode(dev, "p")
+    except:
+        pass
+    try:
+        logging.info("set to high precision")
+        set_mode(dev, "h")
+    except:
+        pass
+    try:
+        logging.info("set to metric mode")
+        set_mode(dev, "m")
+    except:
+        pass
+
+    sch = sched.scheduler(time.time, time.sleep)
+
+    def sample_and_publish_task(scope, delay):
+        sch.enter(delay, 0, sample_and_publish_task, kwargs={
```

Comment:
So I see u have a "timeout" of 3.0 seconds.  And it is possible for us to run this `sample_and_publish_task` function every 1 second.  What happens if the runtime of `sample_and_publish_task` takes longer than 1 second?

would it be better to move this `sch.enter` call to the bottom of the `sample_and_publish_task` function?

Reply:
We talked about this offline, but I'll add a summary of my points here for future use:

1. sched is “single threaded”, so we’ll never have multiple reads at the same time.
2. sched completes the entire function call, no matter how long it takes.
3. Once a called function function returns, the next “ready” task will be run in order of time and priority.
4. I put the enter at the beginning is to “keep the cadence” ignoring small variations in how long the work takes (say 0.3s vs 0.4s), but I’ll think about this more.